src/hmem_cuda_gdrcopy.c

```diff
@@ -0,0 +1,380 @@
+{
+	dlclose(gdrapi_handle);
+	return FI_SUCCESS;
+}
+
+#else
+
+static struct gdrcopy_ops global_gdrcopy_ops = {
+	.gdr_open = gdr_open,
+	.gdr_close = gdr_close,
+	.gdr_pin_buffer = gdr_pin_buffer,
+	.gdr_unpin_buffer = gdr_unpin_buffer,
+	.gdr_map = gdr_map,
+	.gdr_unmap = gdr_unmap,
+	.gdr_copy_to_mapping = gdr_copy_to_mapping,
+	.gdr_copy_from_mapping = gdr_copy_from_mapping
+};
+
+static int cuda_gdrcopy_dl_hmem_init(void)
+{
+	return FI_SUCCESS;
+}
+
+static int cuda_gdrcopy_dl_hmem_cleanup(void)
+{
+	return FI_SUCCESS;
+}
+
+#endif /* ENABLE_CUDA_DLOPEN */
+
+int cuda_gdrcopy_hmem_init(void)
+{
+	int err;
+
+	err = cuda_gdrcopy_dl_hmem_init();
+	if (err) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "gdrcopy_dl_hmem_init failed!\n");
+		return -FI_ENOSYS;
+	}
+
+	assert(global_gdrcopy_ops.gdr_open);
+
+	global_gdr = global_gdrcopy_ops.gdr_open();
+	if (!global_gdr) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "gdr_open failed!\n");
+		cuda_gdrcopy_dl_hmem_cleanup();
+		return -FI_ENOMEM;
+	}
+
+	err = pthread_spin_init(&global_gdr_lock, 0);
+	if (err) {
+		assert(global_gdrcopy_ops.gdr_close);
+		global_gdrcopy_ops.gdr_close(global_gdr);
+		cuda_gdrcopy_dl_hmem_cleanup();
+		return err;
+	}
+
+	return FI_SUCCESS;
+}
+
+int cuda_gdrcopy_hmem_cleanup(void)
+{
+	int err, ret = 0;
+
+	err = pthread_spin_destroy(&global_gdr_lock);
+	if (err) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "destroy global_gdr_lock failed! err: %s\n",
+			strerror(err));
+		ret = err;
+	}
+
+	assert(global_gdrcopy_ops.gdr_close);
+	err = global_gdrcopy_ops.gdr_close(global_gdr);
+	if (err) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "close global_gdr failed! err: %s\n",
+			strerror(err));
+		ret = err;
+	}
+
+	err = cuda_gdrcopy_dl_hmem_cleanup();
+	if (err) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "cuda_gdrcopy_dl_hmem_cleaup() failed! err: %s\n",
+			strerror(err));
+		ret = err;
+	}
+
+	return ret;
+}
+
+int cuda_gdrcopy_to_dev(uint64_t handle, void *devptr, const void *hostptr, size_t len)
+{
+	ssize_t off;
+	struct ofi_gdrcopy_handle *gdrcopy;
+	void *gdrcopy_user_ptr;
+
+	assert(global_gdrcopy_ops.gdr_copy_to_mapping);
+	assert(handle);
+
+	gdrcopy = (struct ofi_gdrcopy_handle *)handle;
+	off = (char *)devptr - (char *)gdrcopy->cuda_ptr;
+	assert(off >= 0 && off + len <= gdrcopy->length);
+	gdrcopy_user_ptr = (char *)gdrcopy->user_ptr + off;
+	global_gdrcopy_ops.gdr_copy_to_mapping(gdrcopy->mh, gdrcopy_user_ptr, hostptr, len);
+	return 0;
+}
+
+int cuda_gdrcopy_from_dev(uint64_t handle, void *hostptr, const void *devptr, size_t len)
+{
+	ssize_t off;
+	struct ofi_gdrcopy_handle *gdrcopy;
+	void *gdrcopy_user_ptr;
+
+	assert(global_gdrcopy_ops.gdr_copy_from_mapping);
+
+	gdrcopy = (struct ofi_gdrcopy_handle *)handle;
+	off = (char *)devptr - (char *)gdrcopy->cuda_ptr;
+	assert(off >= 0 && off + len <= gdrcopy->length);
+	gdrcopy_user_ptr = (char *)gdrcopy->user_ptr + off;
+	global_gdrcopy_ops.gdr_copy_from_mapping(gdrcopy->mh, gdrcopy_user_ptr, hostptr, len);
+	return FI_SUCCESS;
+}
```

Comment:
The above functions could be combined into a single call, with a flag indicating the direction of the copy.

Reply:
Done.